crates/bevy_render/src/mesh/mesh/mod.rs

```diff
@@ -268,24 +268,69 @@ impl Mesh {
     /// Calculates the [`Mesh::ATTRIBUTE_NORMAL`] of a mesh.
     ///
     /// # Panics
-    /// Panics if [`Indices`] are set or [`Mesh::ATTRIBUTE_POSITION`] is not of type `float3`.
-    /// Consider calling [`Mesh::duplicate_vertices`] or export your mesh with normal attributes.
-    pub fn compute_flat_normals(&mut self) {
-        assert!(self.indices().is_none(), "`compute_flat_normals` can't work on indexed geometry. Consider calling `Mesh::duplicate_vertices`.");
+    /// Panics if [`Mesh::ATTRIBUTE_POSITION`] is not of type `float3`.
+    /// Panics if the mesh has any other topology than [`PrimitiveTopology::TriangleList`].
+    ///
+    /// FIXME: The should handle more cases since this is called as a part of gltf
+    /// mesh loading where we can't really blame users for loading meshes that might
+    /// not conform to the limitations here!
+    pub fn compute_normals(&mut self) {
+        assert!(
+            matches!(self.primitive_topology, PrimitiveTopology::TriangleList),
+            "can only compute normals for `TriangleList`s"
+        );
 
         let positions = self
             .attribute(Mesh::ATTRIBUTE_POSITION)
             .unwrap()
             .as_float3()
             .expect("`Mesh::ATTRIBUTE_POSITION` vertex attributes should be of type `float3`");
 
-        let normals: Vec<_> = positions
-            .chunks_exact(3)
-            .map(|p| face_normal(p[0], p[1], p[2]))
-            .flat_map(|normal| [normal; 3])
-            .collect();
-
-        self.set_attribute(Mesh::ATTRIBUTE_NORMAL, normals);
+        match self.indices() {
+            Some(indices) => {
+                let mut count: usize = 0;
+                let mut corners = [0_usize; 3];
+                let mut normals = vec![[0.0f32; 3]; positions.len()];
+                let mut adjacency_counts = vec![0_usize; positions.len()];
+
+                for i in indices.iter() {
+                    corners[count % 3] = i;
+                    count += 1;
+                    if count % 3 == 0 {
+                        let c0 = corners[0];
+                        let c1 = corners[1];
+                        let c2 = corners[2];
+                        let normal = face_normal(positions[c0], positions[c1], positions[c2]);
+                        normals[c0] = (Vec3::from(normal) + Vec3::from(normals[c0])).into();
+                        normals[c1] = (Vec3::from(normal) + Vec3::from(normals[c1])).into();
+                        normals[c2] = (Vec3::from(normal) + Vec3::from(normals[c2])).into();
+                        adjacency_counts[c0] += 1;
+                        adjacency_counts[c1] += 1;
+                        adjacency_counts[c2] += 1;
```

Comment:
nit. there's a fair bit of duplication here, I think a for loop over the corners would remove it and make intent clearer; i.e. we're doing this for every corner
```suggestion
                        let normal = face_normal(
                            positions[corners[0]],
                            positions[corners[1]],
                            positions[corners[2]],
                        );
                        for corner in corners {
                            normals[corner] =
                                (Vec3::from(normal) + Vec3::from(normals[corner])).into();
                            adjacency_counts[corner] += 1;
                        }
```

Reply:
yep, makes sense. not sure why I didn't do that initially